app.py

```diff
@@ -174,47 +143,59 @@ async def update_cl(config: dict, settings: dict):
 # https://docs.chainlit.io/examples/qa
 @cl.on_chat_start
 async def on_chat_start():
-    "Send a chat start message to the chat and load the model config."
+    """
+    Triggered at the start of a chat session. It loads the model configuration from a file
+    and sets it in the user session for future use.
+    """
     config = load_config()
     cl.user_session.set("config", config)
-    await update_cl(config, None)
+    await update_cl(config, {})
 
-    await cl.Message(content=config["introduction"], disable_feedback=True).send()
+    await cl.Message(content=config["introduction"]).send()
 
 
 @cl.on_message
 async def on_message(message: cl.Message):
-    """Handle a message.
+    "Chat message handler."
+    runnable = cl.user_session.get("chain")
+    msg = cl.Message(content="")
+
+    class PostMessageHandler(BaseCallbackHandler):
+        """
+        Callback handler for handling the retriever and LLM processes.
+        Used to post the sources of the retrieved documents as a Chainlit element.
+        """
+
+        def __init__(self, msg: cl.Message):
+            BaseCallbackHandler.__init__(self)
+            self.msg = msg
+            self.sources = set()  # To store unique pairs
+
+        def on_retriever_end(self, documents):
+            "Save the sources found by the retriever."
+            for d in documents:
+                self.sources.add(d.metadata["source"])  # Add unique pairs to the set
+
+        def on_llm_end(self):
+            "Stream the sources as a Chainlit element."
+            if self.sources:
+                sources_text = "\n".join(self.sources)
+                self.msg.elements.append(
+                    cl.Text(name="Sources", content=sources_text, display="inline")
+                )
 
-    Args:
-        message (cl.Message): User prompt input
-    """
-    chain = cl.user_session.get("chain")
-    res = await cl.make_async(chain)(
+    async for chunk in runnable.astream(
```

Comment:
This stream interface is better for the user experience, before there wasn't really any simple options to async stream my runnable.